internal/query/map.go

```diff
@@ -1,81 +1,198 @@
+			if err != nil {
+				allErrors = append(allErrors, err)
+				continue
+			}
+		case filteredRejected:
+			continue
+		case emptyKeyValue:
+			result[key] = value[0]
+		default:
+			err := setValueOnPath(result, []string{key}, value)
 			if err != nil {
+				allErrors = append(allErrors, err)
 				continue
 			}
-			setValueOnPath(result, path, value)
-			exists = true
 		}
 	}
-	if !exists {
-		return nil, exists
+	if len(allErrors) > 0 {
+		return nil, errors.Join(allErrors...)
+	} else if len(result) == 0 {
+		return nil, nil
 	}
-	return result, exists
+	return result, nil
+}
 
+// getType is an internal function to get the type of query key.
+func getType(key string, filteredKey string, getAll bool) queryKeyType {
+	if getAll {
+		if isMap(key) {
+			return mapType
+		} else if isArray(key) {
+			return arrayType
+		} else if key == "" {
+			return emptyKeyValue
+		}
+		return valueType
+	}
+	if isFilteredKey(key, filteredKey) {
+		if isMap(key) {
+			return filteredMap
+		}
+		return filteredUnsupported
+	}
+	return filteredRejected
+}
+
+// isFilteredKey is an internal function to check if k is accepted when searching for map with given key.
+func isFilteredKey(k string, filteredKey string) bool {
+	return k == filteredKey || strings.HasPrefix(k, filteredKey+"[")
+}
+
+// isMap is an internal function to check if k is a map query key.
+func isMap(k string) bool {
+	i := strings.IndexByte(k, '[')
+	j := strings.IndexByte(k, ']')
+	return j-i > 1 || (i >= 0 && j == -1)
 }
 
-// isKey is an internal function to check if a k is a map key.
-func isKey(k string, key string) bool {
+// isArray is an internal function to check if k is an array query key.
+func isArray(k string) bool {
 	i := strings.IndexByte(k, '[')
-	return i >= 1 && k[0:i] == key
+	j := strings.IndexByte(k, ']')
+	return j-i == 1
+}
+
+// keyWithoutArraySymbol is an internal function to remove array symbol from query key.
+func keyWithoutArraySymbol(key string) string {
+	return key[:len(key)-2]
 }
 
 // parsePath is an internal function to parse key access path.
 // For example, key[foo][bar] will be parsed to ["foo", "bar"].
-func parsePath(k string, key string) ([]string, error) {
-	rawPath := strings.TrimPrefix(k, key)
-	if rawPath == "" {
-		return nil, fmt.Errorf("expect %s to be a map but got value", key)
+func parsePath(k string) ([]string, error) {
+	firstKeyEnd := strings.IndexByte(k, '[')
+	if firstKeyEnd == -1 {
+		return nil, fmt.Errorf("invalid access to map key %s", k)
 	}
-	splitted := strings.Split(rawPath, "]")
-	paths := make([]string, 0)
-	for _, p := range splitted {
-		if p == "" {
-			continue
-		}
-		if strings.HasPrefix(p, "[") {
+	first, rawPath := k[:firstKeyEnd], k[firstKeyEnd:]
+
+	split := strings.Split(rawPath, "]")
+
+	// Bear in mind that split of the valid map will always have "" as the last element.
+	if split[len(split)-1] != "" {
+		return nil, fmt.Errorf("invalid access to map key %s", k)
+	} else if len(split)-1 > MaxNestedMapDepth {
```

Comment:
Unnecessary `else` - but it depends on you code style.
- for me it should be removed 🙈